---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

# VSHunter: Capture variation signatures from genomic data

[![GitHub tag](https://img.shields.io/github/tag/ShixiangWang/VSHunter.svg?label=Github)](https://github.com/ShixiangWang/VSHunter) [![Build status](https://ci.appveyor.com/api/projects/status/2n62nnwl8u2dpmc5/branch/master?svg=true)](https://ci.appveyor.com/project/ShixiangWang/vshunter/branch/master)

The goal of VSHunter is to capture variation signature from genomic data. For now, we decode copy number pattern from **absolute copy number profile**. This package collects R code from paper _[Copy number signatures and mutational processes in ovarian carcinoma](https://www.nature.com/articles/s41588-018-0179-8)_ and tidy them as a open source R package for bioinformatics community.

Before you use this tool, you have to obtain **absolute copy number profile** for samples via software like ABSOLUTE v2, QDNASeq etc..

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

## Procedure

1. summarise copy-number profile using a number of different feature distributions:
    * Sgement size
    * Breakpoint number (per ten megabases)
    * change-point copy-number
    * Breakpoint number (per chromosome arm)
    * Length of segments with oscillating copy-number
2. apply mixture modelling to breakdown each feature distribution into mixtures of Gaussian or mixtures of Poisson distributions using the flexmix package.
3. generate a sample-by-component matrix representing the sum of posterior probabilities of each copy-number event being assigned to each component.
4. use NMF package to factorise the sample-by-component matrix into a signature-by-sample matrix and component-by signature-matrix.

```{r, fig.cap="Copy number signature identification, Macintyre, Geoff, et al.(2018)", fig.align='center', eval=TRUE, echo=FALSE}
knitr::include_graphics(path = "https://media.springernature.com/m685/springer-static/image/art%3A10.1038%2Fs41588-018-0179-8/MediaObjects/41588_2018_179_Fig1_HTML.png")
```

```{r, echo=FALSE, include=FALSE, message=FALSE, warning=FALSE}
# devtools::load_all()
# library(flexmix)
# library(QDNAseq)
# library(NMF)
if(!require("VSHunter")){
  devtools::load_all()
}
```

## Installation


You can install UCSCXenaTools from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
devtools::install_github("ShixiangWang/VSHunter", build_vignettes = TRUE)
```


> update features and function will show in vignettes in the future

Load package.

```{r, eval=FALSE}
library(VSHunter)
```



## Citation

* _Macintyre, Geoff, et al. "Copy number signatures and mutational processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262._

If you wanna thank my work for this package, you can also cite (and inlucde link of this package - <https://github.com/ShixiangWang/VSHunter>):

* _Wang, Shixiang, et al. "APOBEC3B and APOBEC mutational signature as potential predictive markers for immunotherapy response in non-small cell lung cancer." Oncogene (2018)._
