---
title: "VSHunter: Decode Pattern of Copy Number Profile"
author: "Shixiang Wang \\
        ShanghaiTech. University"
date: "`r Sys.Date()`"

output:
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
  pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{download publical data in batch quickly and easily}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r, echo=FALSE, include=FALSE, message=FALSE, warning=FALSE}
# devtools::load_all()
# library(flexmix)
# library(QDNAseq)
# library(NMF)
if(!require("VSHunter")){
  devtools::load_all()
}
```

The goal of VSHunter is to capture variation signature from genomic data. For now, we decode copy number pattern from **absolute copy number profile**. This package collects R code from paper _[Copy number signatures and mutational processes in ovarian carcinoma](https://www.nature.com/articles/s41588-018-0179-8)_ and tidy them as a open source R package for bioinformatics community.

Before you use this tool, you have to obtain **absolute copy number profile** for samples via software like ABSOLUTE v2, QDNASeq etc..

## Procedure

1. summarise copy-number profile using a number of different feature distributions:
    * Sgement size
    * Breakpoint number (per ten megabases)
    * change-point copy-number
    * Breakpoint number (per chromosome arm)
    * Length of segments with oscillating copy-number
2. apply mixture modelling to breakdown each feature distribution into mixtures of Gaussian or mixtures of Poisson distributions using the flexmix package.
3. generate a sample-by-component matrix representing the sum of posterior probabilities of each copy-number event being assigned to each component.
4. use NMF package to factorise the sample-by-component matrix into a signature-by-sample matrix and component-by signature-matrix.

```{r, fig.cap="Copy number signature identification, Macintyre, Geoff, et al.(2018)", fig.align='center', eval=TRUE, echo=FALSE}
knitr::include_graphics(path = "https://media.springernature.com/m685/springer-static/image/art%3A10.1038%2Fs41588-018-0179-8/MediaObjects/41588_2018_179_Fig1_HTML.png")
```



## Installation

You can install UCSCXenaTools from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
devtools::install_github("ShixiangWang/VSHunter", build_vignettes = TRUE)
```

Load package.

```{r, eval=FALSE}
library(VSHunter)
```


## Usage

Load example data:

```{r}
load(system.file("extdata/example_cn_list.RData", package = "VSHunter"))
```

`tcga_segTabs` is a list contain absolute copy number profile for multiple samples, each sample is a `data.frame` in the list.

Obtain CNV summary info.


```{r}
tcga_frac = cnv_getLengthFraction(tcga_segTabs)
```


### Derive feature distributions

```{r}
tcga_features = cnv_derivefeatures(CN_data = tcga_segTabs, cores = 1, genome_build = "hg19")
```

### Fit model components

```{r}
tcga_components = cnv_fitMixModels(CN_features = tcga_features, cores = 4)
```


### Generate a sample-by-component matrix

Generate a sample-by-component matrix representing the sum of posterior probabilities of each copy-number event being assigned to each component.

```{r}
tcga_sample_component_matrix = cnv_generateSbCMatrix(tcga_features, tcga_components, cores = 4)
```


### Choose optimal number of signatures.

```{r}
tcga_sig_choose = cnv_chooseSigNumber(tcga_sample_component_matrix, nrun = 10, cores = 4)
```


Do not test a randomise data (save time).

```{r}
tcga_sig_choose2 = cnv_chooseSigNumber(tcga_sample_component_matrix, nrun = 10, cores = 4, testRandom = FALSE)
```


### Extract signatures

```{r}
tcga_signatures = cnv_extractSignatures(tcga_sample_component_matrix, nsig = 3, cores = 4)
```

### Quantify exposure for samples

```{r}
w = NMF::basis(tcga_signatures)
#h = NMF::coef(tcga_signatures)
tcga_exposure = cnv_quantifySigExposure(sample_by_component = tcga_sample_component_matrix, component_by_signature = w)
```

### Auto-capture signatures

Function `cnv_autoCaptureSignatures()` finish three steps (choose best number of signatures, extract signatures and quantify exposure) above in an  antomated way. The arguments of this function are same as `cnv_chooseSigNumber()`.

```{r, eval=FALSE}
tcga_results = cnv_autoCaptureSignatures(tcga_sample_component_matrix, nrun=10, cores = 4)
```

The result object is a `list` which contains all results need fro downstream analysis, include NMF result related to best rank value, signature matrix, absolute and relative exposure (contribution) and best rank survey etc..



### CNV signature calling pipeline

This feature is implemented in `cnv_pipe()` function.

 
 
## Visualization 


Visualize CNV distribution by normalized CN length or chromosome.

```{r}
cnv_plotDistributionProfile(tcga_frac)
```

```{r}
cnv_plotDistributionProfile(tcga_frac, mode = "cd")
```

```{r}
cnv_plotDistributionProfile(tcga_frac, mode = "cd" , fill = TRUE)
```


Plot functions:

```{r, eval=FALSE}
cnv_plotDistributionProfile()
cnv_plotFeatureDistribution()
cnv_plotMixComponents()
cnv_plotSignatures()
```

 
## Citation

* _Macintyre, Geoff, et al. "Copy number signatures and mutational processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262._

If you wanna thank my work for this package, you can also cite (and inlucde link of this package - <https://github.com/ShixiangWang/VSHunter>):

* _Wang, Shixiang, et al. "APOBEC3B and APOBEC mutational signature as potential predictive markers for immunotherapy response in non-small cell lung cancer." Oncogene (2018)._
