<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>cnPattern: Decode Pattern of Copy Number Profile • cnPattern</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="cnPattern: Decode Pattern of Copy Number Profile">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cnPattern</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cnPattern.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ShixiangWang/cnPattern">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>cnPattern: Decode Pattern of Copy Number Profile</h1>
                        <h4 class="author">Shixiang Wang  ShanghaiTech. University</h4>
            
            <h4 class="date">2018-11-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ShixiangWang/cnPattern/blob/master/vignettes/cnPattern.Rmd"><code>vignettes/cnPattern.Rmd</code></a></small>
      <div class="hidden name"><code>cnPattern.Rmd</code></div>

    </div>

    
    
<p>The goal of cnPattern is to decode copy number pattern from <strong>absolute copy number profile</strong>. This package collects R code from paper <em><a href="https://www.nature.com/articles/s41588-018-0179-8">Copy number signatures and mutational processes in ovarian carcinoma</a></em> and tidy them as a open source R package for bioinformatics community.</p>
<p>Before you use this tool, you have to obtain <strong>absolute copy number profile</strong> for samples via software like ABSOLUTE v2, QDNASeq etc..</p>
<div id="procedure" class="section level2">
<h2 class="hasAnchor">
<a href="#procedure" class="anchor"></a>Procedure</h2>
<ol style="list-style-type: decimal">
<li>summarise copy-number profile using a number of different feature distributions:
<ul>
<li>Sgement size</li>
<li>Breakpoint number (per ten megabases)</li>
<li>change-point copy-number</li>
<li>Breakpoint number (per chromosome arm)</li>
<li>Length of segments with oscillating copy-number</li>
</ul>
</li>
<li>apply mixture modelling to breakdown each feature distribution into mixtures of Gaussian or mixtures of Poisson distributions using the flexmix package.</li>
<li>generate a sample-by-component matrix representing the sum of posterior probabilities of each copy-number event being assigned to each component.</li>
<li>use NMF package to factorise the sample-by-component matrix into a signature-by-sample matrix and component-by signature-matrix.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="https://media.springernature.com/m685/springer-static/image/art%3A10.1038%2Fs41588-018-0179-8/MediaObjects/41588_2018_179_Fig1_HTML.png" alt="Copy number signature identification, Macintyre, Geoff, et al.(2018)"><p class="caption">
Copy number signature identification, Macintyre, Geoff, et al.(2018)
</p>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install UCSCXenaTools from github with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"ShixiangWang/cnPattern"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Load package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(cnPattern)</code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>Load example data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="kw">system.file</span>(<span class="st">"inst/extdata"</span>, <span class="st">"example_cn_list.RData"</span>, <span class="dt">package =</span> <span class="st">"cnPattern"</span>))</code></pre></div>
<p><code>tcga_segTabs</code> is a list contain absolute copy number profile for multiple samples, each sample is a <code>data.frame</code> in the list.</p>
<p>Derive feature distributions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tcga_features =<span class="st"> </span><span class="kw"><a href="../reference/derive_features.html">derive_features</a></span>(<span class="dt">CN_data =</span> tcga_segTabs, <span class="dt">cores =</span> <span class="dv">4</span>, <span class="dt">genome_build =</span> <span class="st">"hg19"</span>)</code></pre></div>
<p>Fit model components.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tcga_components =<span class="st"> </span><span class="kw"><a href="../reference/fit_mixModels.html">fit_mixModels</a></span>(<span class="dt">CN_features =</span> tcga_features, <span class="dt">cores =</span> <span class="dv">4</span>)</code></pre></div>
<p>Generate a sample-by-component matrix representing the sum of posterior probabilities of each copy-number event being assigned to each component.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tcga_sample_component_matrix =<span class="st"> </span><span class="kw"><a href="../reference/generate_sbcMatrix.html">generate_sbcMatrix</a></span>(tcga_features, tcga_components, <span class="dt">cores =</span> <span class="dv">4</span>)</code></pre></div>
<p>Choose optimal number of signatures.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tcga_sig_choose =<span class="st"> </span><span class="kw"><a href="../reference/choose_nSignatures.html">choose_nSignatures</a></span>(tcga_sample_component_matrix, <span class="dt">nrun =</span> <span class="dv">10</span>, <span class="dt">cores =</span> <span class="dv">4</span>)</code></pre></div>
<p>Extract signatures.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tcga_signatures =<span class="st"> </span><span class="kw"><a href="../reference/extract_Signatures.html">extract_Signatures</a></span>(tcga_sample_component_matrix, <span class="dt">nsig =</span> <span class="dv">3</span>, <span class="dt">cores =</span> <span class="dv">4</span>)</code></pre></div>
<p>Quantify exposure for samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">w =<span class="st"> </span>NMF<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/NMF/topics/basis-coef-methods">basis</a></span>(tcga_signatures)
<span class="co">#h = NMF::coef(tcga_signatures)</span>
tcga_exposure =<span class="st"> </span><span class="kw"><a href="../reference/quantify_Signatures.html">quantify_Signatures</a></span>(<span class="dt">sample_by_component =</span> tcga_sample_component_matrix, <span class="dt">component_by_signature =</span> w)</code></pre></div>
<p>Function <code><a href="../reference/autoCapture_Signatures.html">autoCapture_Signatures()</a></code> finish three steps (choose best number of signatures, extract signatures and quantify exposure) above in an antomated way. The arguments of this function are same as <code><a href="../reference/choose_nSignatures.html">choose_nSignatures()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tcga_results =<span class="st"> </span><span class="kw"><a href="../reference/autoCapture_Signatures.html">autoCapture_Signatures</a></span>(tcga_sample_component_matrix, <span class="dt">nrun=</span><span class="dv">10</span>, <span class="dt">cores =</span> <span class="dv">4</span>)</code></pre></div>
<p>The result object is a <code>list</code> which contains all results need fro downstream analysis, include NMF result related to best rank value, signature matrix, absolute and relative exposure (contribution) and best rank survey etc..</p>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<ul>
<li><em>Macintyre, Geoff, et al. “Copy number signatures and mutational processes in ovarian carcinoma.” Nature genetics 50.9 (2018): 1262.</em></li>
</ul>
<p>If you wanna thanks for my work of this package, you can also cite:</p>
<ul>
<li><em>Wang, Shixiang, et al. “APOBEC3B and APOBEC mutational signature as potential predictive markers for immunotherapy response in non-small cell lung cancer.” Oncogene (2018).</em></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#procedure">Procedure</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#usage">Usage</a></li>
      <li><a href="#citation">Citation</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Shixiang Wang, Geoffrey Macintyre.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
